tensorpack/tfutils/symbolic_functions.py

```diff
@@ -212,3 +212,142 @@ def saliency_map(output, input, name="saliency_map"):
+
+    Args:
+        l (tf.Tensor): left of image pair
+        r (tf.Tensor): right of image pair
+        y (tf.Tensor): label (1: similar, 0:not similar)
+        margin (float): horizont for negative examples (y==0)
+        extra (bool, optional): return distances for pos and negs
+
+    Returns:
+        tf.Tensor: constrastive_loss, [pos_dist, neg_dist]
+    """
+    with tf.name_scope("constrastive_loss"):
+        y = tf.cast(y, tf.float32)
+
+        diff = l - r
+        # constrative loss L = 0.5 * Y * D^2 + 0.5 * (1-Y) * {max(0, margin - D)}^2
+        delta = tf.reduce_sum(tf.square(diff), 1)
+        delta_sqrt = tf.sqrt(delta + 1e-6)
+
+        match_loss = delta
+        missmatch_loss = tf.square(tf.nn.relu(margin - delta_sqrt))
+
+        loss = tf.reduce_mean(0.5 * (y * match_loss + (1 - y) * missmatch_loss))
+        pos_dist = tf.reduce_mean(y * delta_sqrt, name="pos-dist")
+        neg_dist = tf.reduce_mean((1 - y) * delta_sqrt, name="neg-dist")
+
+        if extra:
+            return loss, pos_dist, neg_dist
+        else:
+            return loss
+
+
+def cosine_loss(l, r, y):
+    """Loss for Siamese networks (cosine version)
+
+    Args:
+        l (tf.Tensor): left of image pair
+        r (tf.Tensor): right of image pair
+        y (tf.Tensor): label (1: similar, 0:not similar)
+
+    Returns:
+        tf.Tensor: cosine-loss as scalar (and) average_pos_loss, average_neg_loss
+    """
+
+    with tf.name_scope("cosine_loss"):
+        y = 2 * tf.cast(y, tf.float32) - 1
+        pred = tf.reduce_sum(l * r, 1) / (tf.squeeze(l2_norm(l) * l2_norm(r)) + 1e-10)
+
+        return tf.nn.l2_loss(y - pred)
+
+
+def triplet_loss(anchor, positive, negative, margin, extra=False):
+    """Loss for Triplet networks as described in the paper:
+    `Deep Metric Learning using Triplet Network <https://arxiv.org/pdf/1412.6622.pdf>` by Hoffer et al.
+
+    Learn embeddings from an anchor point and a similar input (p) as well as a not similar input (n)
+    Loss is: $$\max(0, m + \Vert a-p\Vert^2 - \Vert a-n\Vert^2)$$
+
+    Args:
+        anchor (tf.Tensor): anchor point
+        positive (tf.Tensor): positiv match
+        negative (tf.Tensor): negativ as missmatch
+        margin (float): horizont for negative examples
+        extra (bool, optional): return additional endpoints
+
+    Returns:
+        tf.Tensor: triplet-loss as scalar (and) average_pos_loss, average_neg_loss
+    """
+
+    with tf.name_scope("triplet_loss"):
+        d_pos = tf.reduce_sum(tf.square(anchor - positive), 1)
+        d_neg = tf.reduce_sum(tf.square(anchor - negative), 1)
+
+        loss = tf.reduce_mean(tf.maximum(0., margin + d_pos - d_neg))
+
+        pos_dist = tf.reduce_mean(tf.sqrt(d_pos + 1e-6))
+        neg_dist = tf.reduce_mean(tf.sqrt(d_neg + 1e-6))
+
+        if extra:
+            return loss, pos_dist, neg_dist
+        else:
+            return loss
+
+
+def soft_triplet_loss(a, p, n, eps=1e-6, extra=True):
```

Comment:
It's not consistent when some functions in this file accept eps as an option but some just use a constant.

Also, could you use a more descriptive name rather than a,p,n?